src/PrestaShopBundle/Controller/Admin/Sell/Catalog/Product/SpecificPriceController.php

```diff
@@ -0,0 +1,239 @@
+            'specificPriceForm' => $form->createView(),
+        ]);
+    }
+
+    /**
+     * @AdminSecurity("is_granted('read', request.get('_legacy_controller'))")
+     *
+     * @return Response
+     */
+    public function editAction(Request $request, int $specificPriceId): Response
+    {
+        $form = $this->getFormBuilder()->getFormFor($specificPriceId);
+        $form->handleRequest($request);
+
+        try {
+            $result = $this->getFormHandler()->handleFor($specificPriceId, $form);
+
+            if ($result->isSubmitted() && $result->isValid()) {
+                $this->addFlash('success', $this->trans('Update successful', 'Admin.Notifications.Success'));
+
+                return $this->redirectToRoute('admin_products_specific_prices_edit', [
+                    'specificPriceId' => $specificPriceId,
+                    'liteDisplaying' => $request->query->has('liteDisplaying'),
+                ]);
+            }
+        } catch (Exception $e) {
+            $this->addFlash('error', $this->getErrorMessageForException($e, $this->getErrorMessages()));
+        }
+
+        return $this->render('@PrestaShop/Admin/Sell/Catalog/Product/SpecificPrice/edit.html.twig', [
+            'specificPriceForm' => $form->createView(),
+        ]);
+    }
+
+    /**
+     * Provides error messages for exceptions
+     *
+     * @return array<string, mixed>
+     */
+    private function getErrorMessages(): array
+    {
+        return [
+            SpecificPriceConstraintException::class => [
+                SpecificPriceConstraintException::NOT_UNIQUE_PER_PRODUCT => $this->trans(
+                    'A specific price already exists for these parameters.',
+                    'Admin.Catalog.Notification'
+                ),
+            ],
+        ];
+    }
+
+    /**
+     * @return FormBuilderInterface
+     */
+    private function getFormBuilder(): FormBuilderInterface
+    {
+        return $this->get('prestashop.core.form.identifiable_object.builder.specific_price_form_builder');
+    }
+
+    /**
+     * @return FormHandlerInterface
+     */
+    private function getFormHandler(): FormHandlerInterface
+    {
+        return $this->get('prestashop.core.form.identifiable_object.handler.specific_price_form_handler');
+    }
+
+    /**
+     * @param SpecificPriceList $specificPriceListForEditing
+     *
+     * @return array<int, array<string, mixed>>
+     */
+    private function formatSpecificPricesList(SpecificPriceList $specificPriceListForEditing): array
+    {
+        $list = [];
+        foreach ($specificPriceListForEditing->getSpecificPrices() as $specificPrice) {
+            $list[] = [
+                'id' => $specificPrice->getSpecificPriceId(),
+                'combination' => $specificPrice->getCombinationName() ?: self::UNSPECIFIED_VALUE_FORMAT,
+                'currency' => $specificPrice->getCurrency() ?? $this->trans('All currencies', 'Admin.Global'),
```

Comment:
ok so if we look it from api perspective indeed i should return some nulable scalar values and don't care about their presentation and maybe its not that hard in this specific case, but what about others - like reduction/impact (which representation also depends from its type?), or period? So it means all that formatting stuff goes to javascript, which will become more complex (as everything is couple times more complex in js usually :smile: ). Also do we have CLDR formatting in js? I didn't see any examples of that yet.
TLDR; I don't know how to do it right, as we may not yet have any concrete conventions about such formatting cases.

Reply:
I get that it's more complicated, but we always took the shortcut in Order page to render and format everything in the controller which is also why you can't find many example of CLDR uses but it is available! I think each use case is possible:

- for empty placeholders you would can inject the appropriate placeholder for each column so it doesn't require any advanced JS
- for period (datetime) you can use the default formatting because it can still be parsed and be interpreted in JS if needed, so the current format is fine, however you should always return `from` and `to` even if they are both null (the `Unlimited` placeholder however should be handled in JS)
- for prices you can return the float value and render it with the CLDR js library it's been used in the currency form and a little in order (see examples below)
- for reduction you will have to return both the value and the type I guess

https://github.com/PrestaShop/PrestaShop/blob/1c6f2e701a79f6d0641eb1b31f894e29d1f89a2c/admin-dev/themes/new-theme/js/pages/order/view/order-product-cancel.ts#L59

(This second example is a bit more complex because we dynamically change the specifications to preview the customizations)
https://github.com/PrestaShop/PrestaShop/blob/e160fc92a3267d50b7c72c66ee173a584565314b/admin-dev/themes/new-theme/js/pages/currency/form/components/LanguageList.vue#L102

Reply:
TLDR; it is technically possible to handle this more strictly and let the FO handle the display

But maybe I'm overthinking this, if the API is strictly used by this component only and the data is only injected in the list for display then maybe it's acceptable to preformat the data completely, maybe we should just rename the action so that it reflects it's displayable data (or maybe I'm overthinking again 😅 )